src/repr/type_binary.mli

```diff
@@ -17,6 +17,27 @@
 open Type_core
 open Staging
 
+module Uuid : sig
+  type t = string
+
+  val t : t ty
+  val of_string : string -> t
+end
+
+module Shape : sig
+  type t
+  (** The type of {i shapes} of a particular type with respect to the binary
+      decoders defined here. Shapes are represented canonically such that
+      equality of shapes of two types [t1] and [t2] implies equivalence of the
+      binary codecs derived from [t1] and [t2].
```

Comment:
As the `Shape` module is disjoint from the `Encode` and `Decode` modules, there is no guarantee of this equivalence. There may be accidental discrepancies between the shape and the codecs.

One solution to avoid this problem would be to derive codecs from shapes.

Reply:
Agreed. I'd like to do that, but it has a problem: deriving codecs from shapes requires `Shape.t` to be a GADT w/ witnesses for the algebraic cases – and we don't support type representations for GADTs. This would force us to write out comparisons / pretty-printing / serialisers for `Shape.t` by hand. Those functions would also have to unfold recursive loops (e.g. with the `Unrolling.t` monad), so this starts to look quite a lot like defining a third intermediate typed AST that sits between `Type.t` and (the current) `Shape.t` – I'm not sure it's worth the effort.

Reply:
Maybe if shape was the core of `Type.t` it would be less of a problem to implement, but it might not be worth the effort indeed.